1.0.6 breaks method handling with complex Collection parameter types

Given the following:

```java
import com.google.common.base.Splitter;
import com.google.common.collect.ImmutableList;

import junitparams.JUnitParamsRunner;
import junitparams.Parameters;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Collection;
import java.util.List;

/**
 * Test failing collection usage of parameters annotation.
 */
@RunWith(JUnitParamsRunner.class)
public class BreakerTest {

  @Test
  @Parameters(method = "words")
  public void testBreakage(Collection<String> strings) throws Exception {
    System.out.println(strings);
  }

  private ImmutableList<List<String>> words() {
    return ImmutableList.of(
        Splitter.on(' ').splitToList("Just the place for a Snark!"),
        Splitter.on(' ').splitToList("The Bellman cried, as he carried his crew with care"),
        Splitter.on(' ').splitToList("Supporting each man on the top of the tide,"),
        Splitter.on(' ').splitToList("By a finger entwined in his hair.")
    );
  }
}
```
# 1.0.5

The test runs successfully producing:

```
[Just, the, place, for, a, Snark!]
[The, Bellman, cried,, as, he, carried, his, crew, with, care]
[Supporting, each, man, on, the, top, of, the, tide,]
[By, a, finger, entwined, in, his, hair.]
```

# 1.0.6+

The test fails producing:

```
java.lang.IllegalStateException: While trying to create object of class class java.util.ArrayList could not find constructor with arguments matching (type-wise) the ones given in parameters.

	at junitparams.internal.InvokeParameterisedMethod.createObjectOfExpectedTypeBasedOnParams(InvokeParameterisedMethod.java:85)
	at junitparams.internal.InvokeParameterisedMethod.castParamsFromObjects(InvokeParameterisedMethod.java:67)
	at junitparams.internal.InvokeParameterisedMethod.<init>(InvokeParameterisedMethod.java:38)
	at junitparams.internal.ParameterisedTestClassRunner.buildMethodInvoker(ParameterisedTestClassRunner.java:125)
	at junitparams.internal.ParameterisedTestClassRunner.parameterisedMethodInvoker(ParameterisedTestClassRunner.java:118)
	at junitparams.JUnitParamsRunner.methodInvoker(JUnitParamsRunner.java:448)
	at org.junit.runners.BlockJUnit4ClassRunner.methodBlock(BlockJUnit4ClassRunner.java:273)
	at junitparams.JUnitParamsRunner.runChild(JUnitParamsRunner.java:417)
	at junitparams.JUnitParamsRunner.runChild(JUnitParamsRunner.java:386)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:237)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: java.lang.NoSuchMethodException: java.util.ArrayList.<init>(java.lang.String, java.lang.String, java.lang.String, java.lang.String, java.lang.String, java.lang.String)
	at java.lang.Class.getConstructor0(Class.java:3082)
	at java.lang.Class.getConstructor(Class.java:1825)
	at junitparams.internal.InvokeParameterisedMethod.createObjectOfExpectedTypeBasedOnParams(InvokeParameterisedMethod.java:82)
	... 24 more
```

it should be noted that replacing the words method with 

```
private List<String>[] words() {
    return new List[]{
        Splitter.on(' ').splitToList("Just the place for a Snark!"),
        Splitter.on(' ').splitToList("The Bellman cried, as he carried his crew with care"),
        Splitter.on(' ').splitToList("Supporting each man on the top of the tide,"),
        Splitter.on(' ').splitToList("By a finger entwined in his hair.")
    };
  }
```

still works

--- a/src/main/java/junitparams/internal/parameters/toarray/IterableResultToArray.java
+++ b/src/main/java/junitparams/internal/parameters/toarray/IterableResultToArray.java
@@ -3,10 +3,12 @@ package junitparams.internal.parameters.toarray;
 import java.util.ArrayList;
 import java.util.List;
 
-class IterableResultToArray implements ResultToArray {
+class IterableResultToArray extends SimpleIterableResultToArray {
     private Object result;
 
     IterableResultToArray(Object result) {
+        super(result);
+
         this.result = result;
     }
 
@@ -32,17 +34,6 @@ class IterableResultToArray implements ResultToArray {
             // Eat the exception and keep trying
         }
 
-        try {
-            List<Object[]> res = new ArrayList<Object[]>();
-            for (Object[] paramSet : (Iterable<Object[]>) result)
-                res.add(paramSet);
-            return res.toArray();
-        } catch (ClassCastException e1) {
-            // Iterable with consecutive paramsets, each of one param
-            List<Object> res = new ArrayList<Object>();
-            for (Object param : (Iterable<?>) result)
-                res.add(new Object[]{param});
-            return res.toArray();
-        }
+        return super.convert();
     }
 }

--- a/src/main/java/junitparams/internal/parameters/toarray/ParamsToArrayConverter.java
+++ b/src/main/java/junitparams/internal/parameters/toarray/ParamsToArrayConverter.java
@@ -1,9 +1,9 @@
 package junitparams.internal.parameters.toarray;
 
+import org.junit.runners.model.FrameworkMethod;
+
 import java.util.Arrays;
 import java.util.List;
-
-import org.junit.runners.model.FrameworkMethod;
 
 public class ParamsToArrayConverter {
     private FrameworkMethod frameworkMethod;
@@ -13,6 +13,18 @@ public class ParamsToArrayConverter {
     }
 
     public Object[] convert(Object result) {
+        // handle single iterable parameter result case where result is
+        // assignable to the test method parameter
+        if (frameworkMethod.getMethod().getParameterTypes().length == 1) {
+            Class<?> type = frameworkMethod.getMethod().getParameterTypes()[0];
+            if (type.isAssignableFrom(result.getClass())) {
+                SimpleIterableResultToArray converter = new SimpleIterableResultToArray(result);
+                if (converter.isApplicable()) {
+                    return converter.convert();
+                }
+            }
+        }
+
         List<? extends ResultToArray> converters = Arrays.asList(
                 new ObjectArrayResultToArray(result, frameworkMethod),
                 new IterableResultToArray(result),

--- a/src/main/java/junitparams/internal/parameters/toarray/SimpleIterableResultToArray.java
+++ b/src/main/java/junitparams/internal/parameters/toarray/SimpleIterableResultToArray.java
@@ -1,0 +1,33 @@
+package junitparams.internal.parameters.toarray;
+
+import java.util.ArrayList;
+import java.util.List;
+
+class SimpleIterableResultToArray implements ResultToArray {
+    private final Object result;
+
+    SimpleIterableResultToArray(Object result) {
+        this.result = result;
+    }
+
+    @Override
+    public boolean isApplicable() {
+        return Iterable.class.isAssignableFrom(result.getClass());
+    }
+
+    @Override
+    public Object[] convert() {
+        try {
+            List<Object[]> res = new ArrayList<Object[]>();
+            for (Object[] paramSet : (Iterable<Object[]>) result)
+                res.add(paramSet);
+            return res.toArray();
+        } catch (ClassCastException e1) {
+            // Iterable with consecutive paramsets, each of one param
+            List<Object> res = new ArrayList<Object>();
+            for (Object param : (Iterable<?>) result)
+                res.add(new Object[]{param});
+            return res.toArray();
+        }
+    }
+}

--- a/src/test/java/junitparams/IterableMethodTest.java
+++ b/src/test/java/junitparams/IterableMethodTest.java
@@ -49,4 +49,40 @@ public class IterableMethodTest {
 
         return params;
     }
+
+    @Test
+    @Parameters
+    public void shouldHandleIterableParameters(Iterable<String> a) {
+        assertThat(a)
+                .hasSize(1)
+                .containsOnly("a");
+    }
+
+    public List<List<String>> parametersForShouldHandleIterableParameters() {
+        List<List<String>> params = new ArrayList<List<String>>();
+        List<String> nestedParams = new ArrayList<String>();
+
+        nestedParams.add("a");
+        params.add(nestedParams);
+
+        return params;
+    }
+
+    @Test
+    @Parameters
+    public void shouldHandleIterableParametersArr(Iterable<String> a) {
+        assertThat(a)
+                .hasSize(1)
+                .containsOnly("a");
+    }
+
+    public Object[] parametersForShouldHandleIterableParametersArr() {
+        List<List<String>> params = new ArrayList<List<String>>();
+        List<String> nestedParams = new ArrayList<String>();
+
+        nestedParams.add("a");
+        params.add(nestedParams);
+
+        return params.toArray();
+    }
 }

--- a/src/test/java/junitparams/internal/parameters/toarray/SimpleIterableResultToArrayTest.java
+++ b/src/test/java/junitparams/internal/parameters/toarray/SimpleIterableResultToArrayTest.java
@@ -1,0 +1,81 @@
+package junitparams.internal.parameters.toarray;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Collections;
+
+import static junitparams.JUnitParamsRunner.$;
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+/**
+ * Tests for {@link SimpleIterableResultToArray}.
+ */
+public class SimpleIterableResultToArrayTest {
+
+    private static <T> ArrayList<T> newArrayList(T... args) {
+        ArrayList<T> list = new ArrayList<T>();
+        Collections.addAll(list, args);
+        return list;
+    }
+
+    private ArrayList<String> strings;
+
+    @Before
+    public void setUp() throws Exception {
+        strings = newArrayList("SNARK", "BOOJUM");
+    }
+
+    @Test
+    public void testSimpleIterableObjectArray() throws Exception {
+        Object[] actual = new SimpleIterableResultToArray(
+                newArrayList((Object[]) $("SNARK", "BOOJUM"))
+        ).convert();
+
+        assertThat(actual)
+                .isEqualTo($(
+                        $("SNARK"),
+                        $("BOOJUM")
+                ));
+    }
+
+    @Test
+    public void testSimpleIterable() throws Exception {
+        Object[] actual = new SimpleIterableResultToArray(strings).convert();
+        assertThat(actual)
+                .isEqualTo($(
+                        $("SNARK"),
+                        $("BOOJUM")
+                ));
+    }
+
+    @Test
+    public void testNestedManyIterable() throws Exception {
+        @SuppressWarnings("unchecked") ArrayList<ArrayList<String>> result =
+                newArrayList(strings, strings, strings);
+
+        Object[] actual = new SimpleIterableResultToArray(result).convert();
+        Object[] expected = new Object[]{
+                new Object[]{strings},
+                new Object[]{strings},
+                new Object[]{strings}
+        };
+        assertThat(actual)
+                .isEqualTo(expected);
+
+
+    }
+
+    @Test
+    public void testNestedSingleIterable() throws Exception {
+        @SuppressWarnings("unchecked") ArrayList<ArrayList<String>> result =
+                newArrayList(strings);
+
+        Object[] actual = new SimpleIterableResultToArray(result).convert();
+        Object[] expected = new Object[]{new Object[]{strings}};
+        assertThat(actual)
+                .isEqualTo(expected);
+    }
+}

